cmd/copy.go

```diff
@@ -184,10 +183,48 @@ func (raw rawCopyCmdArgs) cookWithId(jobId common.JobID) (cookedCopyCmdArgs, err
 
 	cooked.fromTo = fromTo
 
-	// copy&transform flags to type-safety
-	cooked.stripTopDir = raw.stripTopDir
+	// Check if source has a trailing wildcard on a URL
+	if fromTo.From().IsRemote() {
+		// Because local already handles wildcards via a list traverser, we should only handle the trailing wildcard --strip-top-dir inference remotely.
+		// To avoid getting trapped by parsing a URL and losing a sense of which *s are real, strip the SAS token in a """unsafe""" way.
+		splitURL := strings.Split(cooked.source, "?")
+
+		// replace %2A with %00 (NULL).
+		// Azure storage doesn't support NULL, so nobody has any reason to ever do this
+		// Thus, %00 is our magic number. Understandably, this is an exception to how we handle wildcards, but this isn't a user-facing exception
+		splitURL[0] = strings.ReplaceAll(splitURL[0], "%2A", "%00")
+
+		sourceURL, err := url.Parse(splitURL[0])
+
+		if err != nil {
+			return cooked, fmt.Errorf("failed to encode %s as URL; %s", strings.ReplaceAll(splitURL[0], "%00", "%2A"), err)
+		}
+
+		// Catch trailing wildcard in object name
+		// Ignore wildcard in container name, as that is handled by initResourceTraverser -> AccountTraverser
+		genericResourceURLParts := common.NewGenericResourceURLParts(*sourceURL, fromTo.From())
+
+		// Infer stripTopDir, trim suffix so we can traverse properly
+		if strings.HasSuffix(genericResourceURLParts.GetObjectName(), "/*") || genericResourceURLParts.GetObjectName() == "*" {
+			genericResourceURLParts.SetObjectName(strings.TrimSuffix(genericResourceURLParts.GetObjectName(), "*"))
+			cooked.stripTopDir = true
+		}
+
+		// Check for other *s, error out and explain the usage
+		if strings.Contains(genericResourceURLParts.GetObjectName(), "*") {
+			return cooked, errors.New("cannot use wildcards in URL except in trailing \"/*\". If you wish to use * in your URL, manually encode it to %2A")
```

Comment:
Maybe change to "cannot use wildcards in _path section of_ URL except in trailing /*".  Just to avoid confusing any pedantic folks who see and, and who know we do support it in the container section.